graphql_compiler/tests/test_ir_generation_errors.py

```diff
@@ -227,11 +229,69 @@ def test_fold_directive_constraints(self):
             }
         }'''
 
+        fold_within_fold = '''{
+            Animal {
+                out_Animal_ParentOf @fold {
+                    out_Animal_FedAt @fold {
+                        uuid @output(out_name: "uuid")
+                    }
+                }
+            }
+        }'''
+
+        optional_within_fold = '''{
+            Animal {
+                out_Animal_ParentOf @fold {
+                    out_Animal_FedAt @optional {
+                        uuid @output(out_name: "uuid")
+                    }
+                }
+            }
+        }'''
+
+        recurse_within_fold = '''{
+            Animal {
+                out_Animal_ParentOf @fold {
+                    in_Animal_ParentOf @recurse(depth: 2) {
+                        uuid @output(out_name: "uuid")
+                    }
+                }
+            }
+        }'''
+
+        output_source_within_fold = '''{
+            Animal {
+                out_Animal_ParentOf @fold {
+                    out_Animal_FedAt @output_source {
+                        uuid @output(out_name: "uuid")
+                    }
+                }
+            }
+        }'''
+
+        multiple_vertex_fields_within_fold = '''{
```

Comment:
Maybe one more variant of this testcase would be nice, where the `@fold` initially goes with one inner traverse, and then forks into two different ones within it. I'm confident that test will pass too, but it would be nice to have it covered explicitly.